Source/WebCore/dom/ElementTextDirection.cpp

```diff
@@ -0,0 +1,325 @@
+        node = NodeTraversal::next(*node, &firstNode);
+    }
+}
+
+static void setElementHasDirAutoFlag(Element& element, TextDirectionality textDirectionality)
+{
+    RefPtr parent = element.parentOrShadowHostElement();
+    element.protectedDocument()->setIsDirAttributeDirty();
+
+    switch (textDirectionality) {
+    case TextDirectionality::LTR:
+    case TextDirectionality::RTL:
+        if (element.selfOrPrecedingNodesAffectDirAuto() || (parent && parent->selfOrPrecedingNodesAffectDirAuto()))
+            setNodeHasDirAutoFlagRecursively(element, false);
+        break;
+
+    case TextDirectionality::Auto:
+        setNodeHasDirAutoFlagRecursively(element, true);
+        break;
+
+    case TextDirectionality::Undefined:
+        if (element.selfOrPrecedingNodesAffectDirAuto() && !(parent && parent->selfOrPrecedingNodesAffectDirAuto()) && !is<HTMLBDIElement>(element))
+            setNodeHasDirAutoFlagRecursively(element, false);
+        break;
+    }
+}
+
+static std::optional<TextDirection> computeTextTextDirection(const String& text)
+{
+    if (auto direction = text.defaultWritingDirection())
+        return *direction == U_LEFT_TO_RIGHT ? TextDirection::LTR : TextDirection::RTL;
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeElementDescendantsAutoTextDirection(const Element& element)
+{
+    for (RefPtr child = element.firstChild(); child; ) {
+        // Specs: Skip bdi, script, style nodes.
+        if (child->hasTagName(bdiTag) || child->hasTagName(scriptTag) || child->hasTagName(styleTag)) {
+            child = NodeTraversal::nextSkippingChildren(*child, &element);
+            continue;
+        }
+
+        if (auto* childElement = dynamicDowncast<Element>(*child)) {
+            // Specs: Skip text form controls
+            // Specs: Skip the element whose dir attribute is not in the undefined state
+            if (childElement->isTextField() || elementHasValidTextDirectionality(*childElement)) {
+                child = NodeTraversal::nextSkippingChildren(*child, &element);
+                continue;
+            }
+        }
+
+        // Specs: If child is a slot element whose root is a shadow root,
+        // then return the directionality of that shadow root's host.
+        if (auto* childSlotElement = dynamicDowncast<HTMLSlotElement>(*child)) {
+            if (auto childHost = childSlotElement->shadowHost())
+                return computeElementTextDirection(*childHost);
+        }
+
+        // Specs: If descendant is a Text node, then continue and its
+        // text directionality is not null, then return it.
+        if (child->isTextNode()) {
+            if (auto direction = computeTextTextDirection(child->textContent(true)))
+                return direction;
+        }
+
+        child = NodeTraversal::next(*child, &element);
+    }
+
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeSlotElementAutoTextDirection(const HTMLSlotElement& slotElement)
```

Comment:
This computes the text direction of slot element so we should call that: computeSlotElementTextDirection

Reply:
This function was renamed to `computeSlotElementTextDirection()`. Similarly `computeTextFormControlElementAutoTextDirection()` was renamed to `computeTextFormControlElementTextDirection()`

Reply:
This is an outdated comment. The functions were renamed to `computeTextDirectionOfSlotElement()` and `computeTextDirectionOfFormControlElement()`.